[ARM] support -mlong-calls -fPIC on arm32 #39970

jxz-hw opened the pull request:
support -mlong-calls -fPIC on arm32
try fixing issuse https://github.com/llvm/llvm-project/issues/39970 and maybe helpful for soving the problem encourted in https://github.com/llvm/llvm-project/pull/142982

jiangxuezhi commented:
e.g.
the command is clang -O2 -mlong-calls -fPIC

the code is:
#include <stdio.h>

int main()
{
    printf("hello");
    return 0;
}

the Relocation section is:
 Offset     Info    Type                Sym. Value  Symbol's Name
00000020  00000503 R_ARM_REL32            00000000   .L.str
00000024  00000760 R_ARM_GOT_PREL         00000000   printf

Relocation section '.rel.ARM.exidx' at offset 0x1a8 contains 1 entry:
 Offset     Info    Type                Sym. Value  Symbol's Name
00000000  0000022a R_ARM_PREL31           00000000   .text

efriedma-quic commented on llvm/test/CodeGen/ARM/long-calls.ll:
```diff
@@ -0,0 +1,32 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=armv7-unknown-linux-gnueabi -relocation-model pic -mattr=+long-calls -o - %s \
+; RUN:    | FileCheck %s
```
Can this be integrated into llvm/test/CodeGen/ARM/subtarget-features-long-calls.ll?

jiangxuezhi replied:
Thank you for the suggestion. I have moved the test case to llvm/test/CodeGen/ARM/subtarget-features-long-calls.ll.

efriedma-quic commented on llvm/lib/Target/ARM/ARMISelLowering.cpp:
```diff
@@ -2815,13 +2815,20 @@ ARMTargetLowering::LowerCall(TargetLowering::CallLoweringInfo &CLI,
+          Callee =
+              DAG.getLoad(PtrVt, dl, DAG.getEntryNode(), Callee,
+                          MachinePointerInfo::getGOT(DAG.getMachineFunction()));
+      } else if (Subtarget->genExecuteOnly()) {
```
Does the execute-only codepath work correctly with PIC?

jiangxuezhi replied:
Thank you for the reminder. Considering that XOM is a security feature which prevents the creation of constant pools in executable memory, and referring to this commit: https://github.com/llvm/llvm-project/commit/52a7dd7d7825b269ae569db0e2bcb193c7698666, it makes sense to first check Subtarget->genExecuteOnly() before evaluating isPositionIndependent().
I have updated the code accordingly. Thanks again for your feedback.

efriedma-quic commented on llvm/lib/Target/ARM/ARMISelLowering.cpp:
```diff
@@ -2839,7 +2846,8 @@ ARMTargetLowering::LowerCall(TargetLowering::CallLoweringInfo &CLI,
+                   dyn_cast<ExternalSymbolSDNode>(Callee)) {
       const char *Sym = S->getSymbol();
 
       if (Subtarget->genExecuteOnly()) {
```
Does the external-symbol codepath work correctly with PIC?  (I think you can trigger this with llvm.memset or something like that.)

jiangxuezhi replied:
Thank you for your suggestion. I have attempted to make the external-symbol code path work with PIC.

During this process, I found that using ARMISD::WrapperPIC caused issues in generating correct SDValues for certain symbols. For example, the following error was triggered:

```
LLVM ERROR: Cannot select: t20: i32 = ARMISD::WrapperPIC TargetExternalSymbol:i32 'memset' [TF=8]
```
This suggests that using the GOT-based approach for PIC is not suitable in this case.

To address this, I instead used the ARMISD::PIC_ADD node to compute PC-relative addresses, which allows proper handling of external symbols under PIC.

In addition, I have added test cases for llvm.memset in: 

llvm-project/llvm/test/CodeGen/ARM/subtarget-features-long-calls.ll

Please help review this change — I'd appreciate your feedback on whether this implementation is appropriate.

efriedma-quic commented on llvm/test/CodeGen/ARM/subtarget-features-long-calls.ll:
```diff
@@ -4,6 +4,9 @@
 ; RUN: llc -mtriple=thumb-- -mcpu=cortex-a8 -relocation-model=static %s -o - -O0 -mattr=-long-calls | FileCheck -check-prefix=NO-LONGCALL %s
+; RUN: llc -mtriple=arm-linux-gnueabi -mcpu=cortex-a8 -relocation-model=pic %s -o - -O0 -mattr=+long-calls | FileCheck -check-prefix=PIC-O0-LONGCALL %s
+; RUN: llc -mtriple=arm-linux-gnueabi -mcpu=cortex-a8 -relocation-model=pic %s -o - -O1 -mattr=+long-calls | FileCheck -check-prefix=PIC-LONGCALL %s
+; RUN: llc -mtriple=arm-linux-gnueabi -mcpu=cortex-a8 -relocation-model=pic %s -o - -O2 -mattr=+long-calls | FileCheck -check-prefix=PIC-LONGCALL %s
```
You don't need both -O1 and -O2; -O0 is special in this context because it triggers globalisel/fastisel.

Please add RUN lines for execute-only, so we can see they work.

Please regenerate the CHECK lines with update_llc_test_checks.py, so it's easier to read/update.

jxz-hw replied:
Done. Removed the extra -O2, added execute-only RUN lines, and regenerated all checks with update_llc_test_checks.py. Should be cleaner now. PTAL, thanks!

efriedma-quic commented on llvm/test/CodeGen/ARM/subtarget-features-long-calls.ll:
```diff
@@ -46,4 +49,62 @@ entry:
+; PIC-LONGCALL-LABEL: test_memset:
+; PIC-LONGCALL:       push {r11, lr}
+; PIC-LONGCALL:       ldr r3, [[MEMSET_LABEL:.*]]
+; PIC-LONGCALL:       add r3, pc, r3
```
The -O1 code is wrong.  The -O0 code is also wrong.

Look at what we do for something like `extern int x; int* f() { return &x; }`.  Or look at what we do for `__attribute((visibility("hidden"))) extern int x; int* f() { return &x; }`.

jxz-hw replied:
Thanks, fixed. ExternalSymbol now uses GOT-based access via a GOT_PREL constant pool entry + PICLDR, matching the existing pattern for non-DSO-local GlobalAddress. Updated the test with dso_local on global_func to separately cover the PC-relative and GOT paths.

efriedma-quic commented on llvm/test/CodeGen/ARM/subtarget-features-long-calls.ll:
```diff
@@ -47,3 +50,80 @@ entry:
+; PIC-XO-LONGCALL-LABEL: test_memset:
+; PIC-XO-LONGCALL:       push {r11, lr}
+; PIC-XO-LONGCALL:       movw r3, :lower16:memset
+; PIC-XO-LONGCALL:       movt r3, :upper16:memset
```
`:lower16:memset` doesn't work with PIC.

jxz-hw replied:
Yeah, sorry for the confusion. The name PIC-XO-LONGCALL was misleading — it made it look like PIC + execute‑only could work together, but they don’t. I’ve fixed the naming and the movw/movt issue. PTAL, thanks!

jxz-hw replied:
Based on my investigation:

ARM's -mexecute-only documentation states that execute-only is not supported for any form of position independent code (PIC), e.g., Armv6-M.
(source : https://developer.arm.com/documentation/101754/0623/armclang-Reference/armclang-Command-line-Options/-mexecute-only)

GCC's -mpure-code (the execute-only alias name) also explicitly does not support -fpic/-fPIC.
(source : https://gcc.gnu.org/legacy-ml/gcc-patches/2019-10/msg01357.html)

LLVM community also explicitly states: "The execute-only and PIC features cannot be enabled together."
(source: https://lists.llvm.org/pipermail/llvm-commits/Week-of-Mon-20170612/460857.html)

Therefore, my commit does not attempt to support the combination of -mlong-calls + -mexecute-only + -fPIC, as it is fundamentally unsupported.

jxz-hw committed:
[ARM] Support -mlong-calls with -fPIC on ARM ELF targets

fix issue #39970 and #142982

- GlobalAddress: use GOT-based addressing via WrapperPIC.  DSO-local symbols use a plain PC-relative WrapperPIC; non-DSO-local symbols additionally load the target address from the GOT.
- ExternalSymbol: load the symbol address from the GOT via a GOT_PREL constant pool entry followed by a PICLDR.
- With -mexecute-only: movw/movt pairs are used instead of constant pools. Note: execute-only may not support PIC; this path is only reached for non-PIC targets.

smithp35 commented on llvm/lib/Target/ARM/ARMISelLowering.cpp:
```diff
@@ -2438,19 +2438,33 @@ ARMTargetLowering::LowerCall(TargetLowering::CallLoweringInfo &CLI,
     if (isa<GlobalAddressSDNode>(Callee)) {
       if (Subtarget->genExecuteOnly()) {
+        // Execute-only forbids constant pools in .text, so use movw/movt.
+        // Note: execute-only may not support fPIC on ARM; this path is only
```
While it is true that fPIC is not implemented correctly when execute-only is enabled, there doesn't seem to be any error message when the two are selected and the function appears to have execute-only attribute.

While I haven't debugged llvm to see if this code path can be hit, I'd like to see some other evidence that genExecuteOnly() is always false when isPositionIndependent() is true.

jxz-hw replied:
Thanks! Good point. I've added an assert to explicitly reject the PositionIndependent + execute-only combination.

smithp35 commented on llvm/lib/Target/ARM/ARMISelLowering.cpp:
```diff
@@ -2438,19 +2438,33 @@ ARMTargetLowering::LowerCall(TargetLowering::CallLoweringInfo &CLI,
+          Callee =
+              DAG.getLoad(PtrVt, dl, DAG.getEntryNode(), Callee,
+                          MachinePointerInfo::getGOT(DAG.getMachineFunction()));
       } else {
```
There is also `isROPI` for embedded which does not use a GOT (no relocations). 
```
        movw    r0, :lower16:(fn-(.LPC2_0+4))
        movt    r0, :upper16:(fn-(.LPC2_0+4))
.LPC2_0:
        add     r0, pc
        ...
```
or with constant pool
```
        ldr     r0, .LCPI2_0
.LPC2_0:
        add     r0, pc
        ...
        .p2align        2
.LCPI2_0:
        .long   baz-(.LPC2_0+4)
```
That sequence would also be more efficient than loading from the GOT for PIC if you happen to know that the destination is non-preemptible.

jxz-hw replied:
Thank you for the reminder. I checked the definition in the LLVM code – according to (https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/TargetMachine.cpp#L144), 
~~~cpp
bool TargetMachine::isPositionIndependent() const {
  return getRelocationModel() == Reloc::PIC_;
}
~~~
and https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/ARM/ARMSubtarget.cpp#L381
~~~cpp
bool ARMSubtarget::isROPI() const {
  // FIXME: This should ideally come from a function attribute, to work
  // correctly with LTO.
  return TM.getRelocationModel() == Reloc::ROPI ||
         TM.getRelocationModel() == Reloc::ROPI_RWPI;
}
~~~

isPositionIndependent() corresponds to PIC_, and Reloc::Model includes:
~~~cpp
namespace Reloc {
  enum Model { Static, PIC_, DynamicNoPIC, ROPI, RWPI, ROPI_RWPI };
}
~~~

Thus, I think isPositionIndependent() and isROPI() are coordinate concepts, so there maybe no case where isROPI() is true but the GOT is mistakenly used. Moreover, both before and after my change, when isROPI() is true, the code consistently uses an absolute address constant pool (PCAdj = 0).

Let me write an example to verify this.

jxz-hw replied:
Options：-mlong-calls -fropi

Example A：Mainly tests calling an external function

~~~c
extern void external_func(void);
void call_external(void) {
    external_func();
}
~~~

The output is the same both before and after my change.
~~~c
	ldr	r0, .LCPI0_0
	bx	r0
.LCPI0_0:
	.long	external_func
~~~
Through GDB debugging, since both Subtarget->genExecuteOnly() and isPIC are false, this primarily corresponds to the following logic in the LLVM code:

~~~cpp
      } else {
        // Neither execute-only nor PIC: load the address from a constant pool.
        unsigned ARMPCLabelIndex = AFI->createPICLabelUId();
        ARMConstantPoolValue *CPV = ARMConstantPoolConstant::Create(
            GVal, ARMPCLabelIndex, ARMCP::CPValue, 0);

        // Get the address of the callee into a register
        SDValue Addr = DAG.getTargetConstantPool(CPV, PtrVt, Align(4));
        Addr = DAG.getNode(ARMISD::Wrapper, dl, MVT::i32, Addr);
        Callee = DAG.getLoad(
            PtrVt, dl, DAG.getEntryNode(), Addr,
            MachinePointerInfo::getConstantPool(DAG.getMachineFunction()));
      }
~~~

Since PCAdj = 0 is passed when creating the ARMConstantPoolConstant, an absolute address load is used.

jxz-hw replied:
Sorry, due to my limited understanding of this area, I would like to ask: for the above use case, is it necessary to support isROPI() under -mlong-calls to avoid the absolute address loading issue? 
Additionally, the interaction between isROPI() and XOM (Execute‑Only Memory) may also need to be considered.

I‘ve add assertion to reject unsupported ROPI + long-calls combination.

~~~c
    assert(!Subtarget->isROPI() &&
           "long-calls with ROPI is not currently supported");
~~~


Alternatively, I can file a separate issue to track this matter, as adapting -mlong-calls -fropi seems to be beyond the scope of fixing the current issue https://github.com/llvm/llvm-project/issues/39970 and https://github.com/llvm/llvm-project/pull/142982.

jxz-hw replied:
Thank you once more for taking the time to review this and for your valuable comments. @smithp35

smithp35 replied:
It isn't necessary to support ROPI code generation as it didn't work beforehand. It would be good to give a user visible error message that ROPI is not compatible with the option though.

I also mentioned it as the code ROPI uses to access variables could be adapted to call non-preemptible functions without needing a GOT access. That is an optimisation though.

I think we're running into a case where the original implementation of `-mlong-calls` has some problems, like producing non-XO or non position independent code when `-mexecute-only` or `-fropi` is used. Adding `-fPIC` support doesn't make those limitations any worse, but it may mean more people use the option and run into these problems. While I don't think support needs to be added, we should error out if we know they are going to generate incorrect code.

jxz-hw replied:
Thanks for your explanation. Yes, I agree with your point, and I have added an assertion to give a user-visible error when -mlong-calls and -fropi are used together, as that combination may generate incorrect code:

~~~c
assert(!Subtarget->isROPI() &&
       "long-calls with ROPI is not currently supported");
~~~

At the same time, the combination of -mlong-calls and -fpic appears to be reasonable – at least GCC already supports the same option pair. Our patch simply helps LLVM catch up in this area.

Thank you for your review. Could you please let me know if this merge request is ready to be merged, or are there any other issues we should discuss further? Thanks. @smithp35

smithp35 replied:
An assert isn't user-visible, as most users will be using a release build with no assertions.

I think the idea of an assert in the backend is that the front-end, which can give much better error messages does not generate LLVM-IR with these combinations.

For me, I would like to see a clang error message when there's a combination of -mlong-calls and execute-only (or ropi). There's also a per function attribute that would need to be checked https://clang.llvm.org/docs/AttributeReference.html#long-call-far  

Other than that, I'm OK when efreidma-quic, or possibly another backend maintainer, is happy with the code-generation.

jxz-hw replied:
Thank you for the reminder. I've now switched to using report_fatal_error instead of an assertion to warn users that these options cannot be combined in the backend. 

~~~cpp
    if (isPIC && Subtarget->genExecuteOnly())
      report_fatal_error(Twine("long-calls with execute-only and "
                               "position-independent code is not supported"),
                         /*gen_crash_diag=*/false);
    if (Subtarget->isROPI())
      report_fatal_error(
          Twine("long-calls with ROPI is not currently supported"),
          /*gen_crash_diag=*/false);
~~~

I also agree that the warning should be emitted at the clang rather than in the backend, and that the attribute should be checked as well. I've created a new commit to address this. Please review, thanks!

Thank you for your suggestion—I've learned a lot and will pay closer attention to these details in my future code contributions.

smithp35 commented on llvm/lib/Target/ARM/ARMISelLowering.cpp:
```diff
@@ -2466,12 +2480,32 @@ ARMTargetLowering::LowerCall(TargetLowering::CallLoweringInfo &CLI,
 
       if (Subtarget->genExecuteOnly()) {
+        // Execute-only forbids constant pools in .text, so use movw/movt.
+        // Note: execute-only may not support fPIC on ARM; this path is only
```
Same comment as above, while fPIC and execute-only doesn't mix, I don't think there is any error message to stop it.

jxz-hw replied:
ok,  I've fixed it

smithp35 reviewed (commented):
Some comments on the history of XO. I don't want to get too deeply involved with the technical details of the review as this isn't my strongest area. It may be that some more error messages need to be added for incompatibile options before we can rule out code-paths.

Arm's position is that range-extension is best done by the linker, so it hasn't officially supported `-mlong-calls` which includes testing or documenting it in any of its official toolchains.

> ARM's -mexecute-only documentation states that execute-only is not supported for any form of position independent code (PIC), e.g., Armv6-M.
(source : https://developer.arm.com/documentation/101754/0623/armclang-Reference/armclang-Command-line-Options/-mexecute-only)

That is the documentation for Arm's proprietary toolchain, which is a downstream fork of LLVM with its own Driver. There are downstream changes for XO and alternative error message. We shouldn't treat this as a reference for how the upstream clang driver behaves.

It is true that XO support was implemented for M-profile as there was demand for it from that market to hide on chip firmware, first for v7-M then later for v6-M (even though the code generation is truly horrible). There is nothing fundamental that prevents its use on other profiles, it is just that there hasn't been demand for it. Nor have all the edge cases been tested.

While I think it is theoretically possible to implement PIC with XO, the code-generation wouldn't be great as in the current ABI the GOT relative relocations only operate on data so the constant-pools containing offsets into the GOT would need to be moved to a separate non-executable section.

> GCC's -mpure-code (the execute-only alias name) also explicitly does not support -fpic/-fPIC.
(source : https://gcc.gnu.org/legacy-ml/gcc-patches/2019-10/msg01357.html)

GCC has an error message when pure-code and pic are mixed:

> cc1: error: -mpure-code only supports non-pic code on M-profile targets
cc1: error: -mpure-code incompatible with '-mword-relocations'

If we had a similar error message then some of the comments about not reaching the code-path would be stronger.

> LLVM community also explicitly states: "The execute-only and PIC features cannot be enabled together."
(source: https://lists.llvm.org/pipermail/llvm-commits/Week-of-Mon-20170612/460857.html)

That may have been true at one point, but it is certainly possible to enable them both, and PIC seems to silently be used in preference.

> Therefore, my commit does not attempt to support the combination of -mlong-calls + -mexecute-only + -fPIC, as it is fundamentally unsupported.

I agree that it doesn't need to support it, although at the moment it looks like it it is possible to enable both options and end up with broken output. Perhaps a separate patch is needed to guard what is and isn't supported.

It also looks like ROPI (-fropi) would produce incorrect code as that cannot use the GOT.

efriedma-quic commented on llvm/lib/Target/ARM/ARMISelLowering.cpp:
```diff
@@ -2438,19 +2438,40 @@ ARMTargetLowering::LowerCall(TargetLowering::CallLoweringInfo &CLI,
-           "long-calls codegen is not position independent!");
+    bool isPIC = isPositionIndependent() && !TT.isOSWindows();
+    if (isPIC && Subtarget->genExecuteOnly())
+      report_fatal_error(Twine("long-calls with execute-only and "
```
reportFatalUsageError

jxz-hw replied:
Thanks for your reminder, I have fixed it.  report_fatal_error has been marked deprecated.

efriedma-quic commented on llvm/lib/Target/ARM/ARMISelLowering.cpp:
```diff
@@ -2438,19 +2438,40 @@ ARMTargetLowering::LowerCall(TargetLowering::CallLoweringInfo &CLI,
-           "long-calls codegen is not position independent!");
+    bool isPIC = isPositionIndependent() && !TT.isOSWindows();
+    if (isPIC && Subtarget->genExecuteOnly())
+      report_fatal_error(Twine("long-calls with execute-only and "
```
reportFatalUsageError

jxz-hw replied:
already fixed it.

efriedma-quic commented on clang/lib/CodeGen/Targets/ARM.cpp:
```diff
@@ -173,6 +173,37 @@ class ARMTargetCodeGenInfo : public TargetCodeGenInfo {
 
+    if (FD) {
+      bool HasLongCalls = false, HasExecOnly = false;
+      if (Fn->hasFnAttribute("target-features")) {
```
This is pretty complicated code to catch an edge case that won't come up in practice, and the backend will correctly diagnose.  I don't think it's necessary.

jxz-hw replied:
Agreed. The backend can diagnose it, so I've reverted the change.

efriedma-quic commented on llvm/test/CodeGen/ARM/subtarget-features-long-calls.ll:
```diff
@@ -47,3 +130,127 @@ entry:
+; PIC-LONGCALL-NEXT:    .p2align 2
+; PIC-LONGCALL-NEXT:  @ %bb.1:
+; PIC-LONGCALL-NEXT:  .LCPI4_0:
+; PIC-LONGCALL-NEXT:    .long memset(GOT_PREL)
```
Does this need to be `.long memset(GOT_PREL)-(.LPC4_0+8-.Ltmp1)` or something like that?

jxz-hw replied:
Thanks, I fixed it. Now the external symbol and global address both use the same GOT_PREL form.

efriedma-quic reviewed (approved):
LGTM

hstk30-hw reviewed (approved):
Thinks fix it. LGTM
